src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -244,16 +218,115 @@ Status RedisStoreClient::AsyncMultiGet(
   RAY_CHECK(callback);
   if (keys.empty()) {
     callback({});
+    return Status::OK();
   }
   std::vector<std::string> true_keys;
   for (auto &key : keys) {
     true_keys.push_back(GenRedisKey(external_storage_namespace_, table_name, key));
   }
-  RAY_CHECK_OK(MGetValues(
-      redis_client_, external_storage_namespace_, table_name, true_keys, callback));
+  MGetValues(table_name, true_keys, callback);
   return Status::OK();
 }
 
+size_t RedisStoreClient::PushToSendingQueue(const std::vector<std::string> &keys,
+                                            std::function<void()> send_request) {
+  size_t queue_added = 0;
+  for (const auto &key : keys) {
+    auto [op_iter, added] =
+        pending_redis_request_by_key_.emplace(key, std::queue<std::function<void()>>());
+    if (added) {
+      queue_added++;
+    }
+    if (added) {
+      // There is no in-flight requests for this key which means
+      // in the callback the queue won't be poped. So just push
+      // an entry as placeholder to indicate there is pending
+      // requests.
+      op_iter->second.push(nullptr);
+    } else {
+      op_iter->second.push(send_request);
+    }
+  }
+  return queue_added;
+}
+
+std::vector<std::function<void()>> RedisStoreClient::PopFromSendingQueue(
+    const std::vector<std::string> &keys) {
+  std::vector<std::function<void()>> send_requests;
+  for (const auto &key : keys) {
+    auto [op_iter, added] =
+        pending_redis_request_by_key_.emplace(key, std::queue<std::function<void()>>());
+    RAY_CHECK(added == false) << "Pop from a queue doesn't exist: " << key;
+    op_iter->second.pop();
+    if (op_iter->second.empty()) {
+      pending_redis_request_by_key_.erase(op_iter);
+    } else {
+      send_requests.emplace_back(std::move(op_iter->second.front()));
+    }
+  }
+  return send_requests;
+}
+
+void RedisStoreClient::SendRedisCmd(std::vector<std::string> keys,
+                                    std::vector<std::string> args,
+                                    RedisCallback redis_callback) {
+  RAY_CHECK(!keys.empty());
+  // The number of keys that's ready for this request.
+  // For a query reading or writing multiple keys, we need a counter
+  // to check whether all existing requests for this keys have been
+  // processed.
+  auto num_ready_keys = std::make_shared<std::atomic<size_t>>(0);
```

Comment:
nit: looks std::atomic is not needed.

Reply:
The issue here is that we never know where the callback is going to be executed. 
In the redis client context, callback is going to be executed, and the sender might send request in another thread.